### ISSUE TITLE
AM_PROG_LIBTOOL not found on Linux

### ISSUE THREAD

**Issue by guru-florida** [json-c/json-c#308]
./autogen.sh or "autoreconf -fvi" fails with:
configure.ac:78: warning: macro 'AM_PROG_LIBTOOL' not found in library

I have libtool, autoconf and automake installed. Tried both 2.63 and 2.69 (latest) versions of autoconf. Libtool version is 2.2.6b, automake is 1.15. 

CMake build is also broken as it just calls autoconf. Why not just switch to CMake? It isnt that hard to setup and I find real cmake builds always work and autoconf is usually a PITA. I think autoconf fails mostly due to due to its dependencies or perhaps it's own propensity to change/deprecate interfaces and macros. I regularly build and create cmake projects on Mac, Windows, Solaris and Linux and am often surprised at how often the projects are portable out of the gate. :)

**Comment by hawicz** [comment]
hmm.. maybe a difference between automake 1.14 and 1.15?
If you do a "grep -rl AM_PROG_LIBTOOL /usr/share/a*", what do you get?

Try changing the AM_PROG_LIBTOOL line in configure.ac to say LT_INIT instead.

**Comment by guru-florida** [comment]
Thanks for the fast response. I did but it complained about some other LT_INIT related stuff then. As I said, autoconf seems to change stuff all the time making portability difficult IMO. I went ahead and forked your project and fixed up your CMakeLists.txt to not require autoconf/make/libtool. That meant doing all the platform checks using cmake stuff...works similar to how autoconf does it. Still have a config.h.in with the HAVE_xxx_H stuff, and a json_config.h.in as well, but to keep from messing up autoconf I have those two files in /cmake/ directory. Thus autoconf should still work too.

 Check out my fork at...if you want I can create a pull request to get it back in your repo.
https://github.com/guru-florida/json-c

Just do a git clone, and cd into dir, then:
mkdir build
cd build
cmake ..
make
make install

Also:  The cmake build creates both .a and .so files plus with the version numbers on the so file (soft links). Version settings are in the CMakeLists.txt, I could generate the json_c_verson.h file but to do so I might have to break autoconf. "make install" now only installs the headers and libs, doesnt pick up other stuff.

Though I havent checked yet, this should build on windows, mac and linux no problem. I will test soon.

**Comment by hawicz** [comment]
Switching to cmake sounds like it would be reasonable, but I'd like to fix the autoconf problem first.  What platform are you building on when you get the error?

**Comment by hawicz** [comment]
Does cmake have a defined way to run tests?  With your changes so far "make check" no longer work, so I'm guessing there's some other way that's intended to work.

**Comment by guru-florida** [comment]
Yes, cmake supports tests using CTest. I didnt see the tests in the json-c
src so I didnt add that. It also supports making packages via CPack. CPack
can make debian/ubuntu (deb) and RPMs and tgz bin or sources. Adding
support for packaging is pretty simple as most of the info is already set
with the cmake build directives. I just learned cmake last year and I am
hooked...I learned it after noticing that every time I want to build deps
for something cmake build systems always worked and rarely required any
hacking or twiddling of configure options...it just worked and I like that!
:)   I converted over a huge project from Makefiles to cmake and easily got
it to compile on Solaris and Mac where it was just LInux before. On a lark,
I compiled it on Windows under MinGW and under *Windows10 Ubuntu subsystem*
 and it worked no problem. I've also had other cmake projects since that
target embedded systems, various arch and different endians.

Also, I have your json-c project working with CLion IDE (by Jetbrains).
CLion requires using cmake though. You should check it out though, just
download and open the root directory (it picks up the
CMakeLists.txt)...this IDE beats MS Visual Studio! (which was my favorite
IDE before this.) Full code inspection, auto-complete and refactoring
tools. :)  Eclipse for C/C++ is garbage unfortunately so I got used to VIM
on Linux.

I was compiling json-c on CentOS linux 5.7 x86_64.

*Colin MacKenzie*

*FlyingEinstein.com <http://flyingeinstein.com/>*

**Comment by guru-florida** [comment]
I suppose I should get doxygen to also work via cmake as well.

*Colin MacKenzie*

*FlyingEinstein.com <http://flyingeinstein.com/>*

**Comment by hawicz** [comment]
There don't seem to be pre-built packages for that new of versions of autoconf and automake available in CentOS 5, so I'm assuming you're building from source.  When building from source (on CentOS 5.11, but s/b minimally different from 5.7), attempting to install automake-1.15 with autoconf-2.63 fails because it needs a newer version of autoconf.  If you reversed the order and installed autoconf after automake, then perhaps that is the source of some of the problems.
The only way I ran into the problem with AM_PROG_LIBTOOL was when I forgot to run "make install" in the libtool build (after installing autoconf and automake).

The steps I used to get a working (though noisy) set of autotools are:

```set -e
tar xzf autoconf-2.69.tar.gz
tar xzf automake-1.15.tar.gz
tar xzf libtool-2.2.6b.tar.gz

export PATH=${HOME}/ac_install/bin:$PATH

(cd autoconf-2.69 && \
  ./configure --prefix ${HOME}/ac_install && \
  make && \
  make install)

(cd automake-1.15 && \
  ./configure --prefix ${HOME}/ac_install && \
  make && \
  make install)

(cd libtool-2.2.6b && \
  ./configure --prefix ${HOME}/ac_install && \
  make && \
  make install)
```

**Comment by liudf0716** [comment]
It seems most of us encounter such annoying problem when compiling json-c

**Comment by hawicz** [comment]
Note that on a more up to date OS version, it is easier.  For instance, I just tried on on a fresh Ubuntu 16.04.2 LTS install, and the *full* set of steps are:
* sudo apt install git
* sudo apt install autoconf automake libtool
* sudo apt install valgrind  # optional
* git clone https://github.com/json-c/json-c
* cd json-c
* sh autogen.sh
* ./configure
* make && make check && make install

@liudf0716, if neither a series of steps like those, nor the manual build instructions I posted above work, then I find it hard to consider json-c to be at fault, and any difficulty you're running into is a quirk of your environment.  Vague complaints are unhelpful, but if you have specific changes to improve the way json-c builds, then by all means pass the along (ideally in the form of a pull request).
